### PR TITLE
Geolocation improvements, refs #10330

### PR DESCRIPTION
**Pull request by mcantelon**
Redmine: https://projects.artefactual.com/issues/10330

Added ability to specify latitude/longitude of digital object and show resulting map on description pages:
- Add ability for admins to toggle display of digital object map on default page elements page
- Add ability to specify latitude/longitude when editing digital object metadata

AtoM tweaks/fixes related to feature:
- Change Google Maps API setting from config file var to DB setting
- Fix issue with use of get_partial instead of get_component when displaying digital object metadata
- Darken links within info alerts

### CODE REVIEW

**Comment by sevein** [comment]
Ready for review or still work in progress, @mcantelon?

**Comment by mcantelon** [comment]
I talked with Sara and realized that I'd missed one feature... will get PR review ready today.

**Comment by mcantelon** [comment]
Alright, this one's ready to roll.

File: apps/qubit/modules/settings/templates/pageElementsSuccess.php [hunk 1]
```diff
@@ -50,6 +50,18 @@
             <td><?php echo $form->toggleIoSlider ?></td>
           </tr>
           <tr>
+            <td>
+              <?php echo $form->toggleDigitalObjectMap->label('Digital object map')->renderLabel() ?>
+              <?php if (!$googleMapsApiKeySet): ?>
+                <div id="notice-alerts" class="alert alert-info">
```
Comment (sevein) [inline apps/qubit/modules/settings/templates/pageElementsSuccess.php]: Should we use our standard `<div class="description">...</div>` instead? See: https://imgur.com/a/5hH4r.

File: apps/qubit/modules/settings/actions/securityAction.class.php [hunk 2]
```diff
@@ -62,11 +62,13 @@ protected function populateSecurityForm()
     $limitAdminIp = QubitSetting::getByName('limit_admin_ip');
     $requireSslAdmin = QubitSetting::getByName('require_ssl_admin');
     $requireStrongPasswords = QubitSetting::getByName('require_strong_passwords');
+    $googleMapsApiKey = QubitSetting::getByName('google_maps_api_key');
```
Comment (sevein) [inline apps/qubit/modules/settings/actions/securityAction.class.php]: I'm not sure if the security page is the best place for the Google Maps JavaScript API key. It's not even a secret, is it? I'd personally prefer to see it (by order of preference): 1) in the same place where you enable the map element (under default page elements, see: https://imgur.com/a/M0tqq), or 2) in a new settings section ("Third-party integrations"?), or 3) under global, which is kind of a sink of all kind of settings.

File: plugins/sfIsdiahPlugin/modules/sfIsdiahPlugin/templates/indexSuccess.php [hunk 3]
```diff
@@ -73,8 +73,8 @@
 
 <?php end_slot() ?>
 
-<?php if (isset($latitude) && isset($longitude) && null !== $key = sfConfig::get('app_google_maps_api_key')): ?>
-  <div id="front-map" class="simple-map" data-key="<?php echo $key ?>" data-latitude="<?php echo $latitude ?>" data-longitude="<?php echo $longitude ?>"></div>
+<?php if (isset($latitude) && isset($longitude) && null !== $keyProperty = QubitSetting::getByName('google_maps_api_key')): ?>
+  <div id="front-map" class="simple-map" data-key="<?php echo $keyProperty->value ?>" data-latitude="<?php echo $latitude ?>" data-longitude="<?php echo $longitude ?>"></div>
```
Comment (sevein) [inline plugins/sfIsdiahPlugin/modules/sfIsdiahPlugin/templates/indexSuccess.php]: This is correct but is it necessary?  `lib/filter/QubitSettingsFilter.class.php` loads sfConfig with the settings in `QubitSetting`. In theory `sfConfig::get('app_google_maps_api_key')` should still return the value of `google_maps_api_key` even after the migration to `QubitSetting`. Not 100% sure but give it a shot? Take a look at the filter too if you can!

File: apps/qubit/modules/digitalobject/actions/metadataComponent.class.php [hunk 4]
```diff
@@ -40,5 +40,17 @@ public function execute($request)
     {
       $this->denyFileNameByPremis = true;
     }
+
+    // Provide Google Maps API key to template
+    $keyProperty = QubitSetting::getByName('google_maps_api_key');
```
Comment (sevein) [inline apps/qubit/modules/digitalobject/actions/metadataComponent.class.php]: If you use the sfConfig store instead you'll save a hit to the database,  settings are populated in `sfConfig` from our cache backend.

File: plugins/sfIsadPlugin/modules/sfIsadPlugin/templates/indexSuccess.php [hunk 5]
```diff
@@ -326,7 +326,7 @@
 <?php if (0 < count($resource->digitalObjects)): ?>
 
   <div class="digitalObjectMetadata">
-    <?php echo get_partial('digitalobject/metadata', array('resource' => $resource->digitalObjects[0])) ?>
+    <?php echo get_component('digitalobject', 'metadata', array('resource' => $resource->digitalObjects[0], 'infoObj' => $resource)) ?>
```
Comment (sevein) [inline plugins/sfIsadPlugin/modules/sfIsadPlugin/templates/indexSuccess.php:329]: Good catch!

**Review by sevein — CHANGES_REQUESTED** [review]
Pretty good, only a few changes suggested!

**Comment by sevein** [comment]
What's the public ticket? #10232 is private, we should name branches after the public ticket id right?

**Comment by sevein** [comment]
Remember to name the branch and the pull request after the public Redmine ticket next time pleaseeee... 🙏

File: apps/qubit/modules/settings/actions/securityAction.class.php (hunk 2 above)
Comment (mcantelon) [reply apps/qubit/modules/settings/actions/securityAction.class.php]: Hrm... maybe we can put it into global until there are other API keys to group with it.

File: plugins/sfIsdiahPlugin/modules/sfIsdiahPlugin/templates/indexSuccess.php (hunk 3 above)
Comment (mcantelon) [reply plugins/sfIsdiahPlugin/modules/sfIsdiahPlugin/templates/indexSuccess.php]: Sounds good.

File: apps/qubit/modules/settings/actions/securityAction.class.php (hunk 2 above)
Comment (sevein) [reply apps/qubit/modules/settings/actions/securityAction.class.php]: Makes sense!

**Comment by mcantelon** [comment]
Oops... PR and commits reference public ticket, but I forgot to rename the branch. Sorry. :(

**Comment by mcantelon** [comment]
OK, ready for another review.

**Comment by sevein** [comment]
Can we merge the translations? Their classes also have the same names?

**Comment by mcantelon** [comment]
I've never dealt with translations so not sure what you mean.

**Comment by sevein** [comment]
Sorry, I meant the database migrations.

**Comment by sevein** [comment]
This is the reason of my comment:
`arMigration0146.class.php` => `class arMigration0146`
`arMigration0147.class.php` => `class arMigration0146`

In any case, I understand that a migration is necessary for `digital_object_geolocation` because we want it to be enabled by default. However, do we need a migration for `google_maps_api_key` knowing that the default is to **not** have a key?

**Comment by mcantelon** [comment]
Oops... sorry I'll fix that.

**Comment by mcantelon** [comment]
OK, fixed!

File: lib/task/migrate/migrations/arMigration0146.class.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    if (null === QubitSetting::getByName('google_maps_api_key'))
+    {
+      $setting = new QubitSetting;
+      $setting->setName('google_maps_api_key');
+      $setting->save();
+    }
+
+    return true;
+  }
+}
```
Comment (sevein) [inline lib/task/migrate/migrations/arMigration0146.class.php:50]: Do we need this migration?

File: data/fixtures/settings.yml [hunk 7]
```diff
@@ -779,6 +779,9 @@ QubitSetting:
   QubitSetting_swordDepositDir:
     name: sword_deposit_dir
     value: /tmp
+  QubitSetting_googleMapsApiKey:
+    name: google_maps_api_key
+    value:
```
Comment (sevein) [inline data/fixtures/settings.yml]: Do we need this?

File: apps/qubit/modules/settings/actions/globalAction.class.php [hunk 8]
```diff
@@ -364,6 +366,16 @@ protected function updateGlobalSettings()
       $setting->save();
     }
 
+    // Google Maps Javascript API key
+    if (null !== $googleMapsApiKey = $thisForm->getValue('google_maps_api_key'))
+    {
+      $setting = QubitSetting::getByName('google_maps_api_key');
```
Comment (sevein) [inline apps/qubit/modules/settings/actions/globalAction.class.php]: We can save the migration for `google_maps_api_key` if you create the object on demand, right here, if `$setting` is `NULL`. You did something very similar in the following commit a while ago: f3041d0d291437d5a0816f8c3488b36661e5f1f2.

**Review by sevein — CHANGES_REQUESTED** [review]
I'm still hoping that we can save that extra migration, can we? Me 15 mins ago:

> In any case, I understand that a migration is necessary for digital_object_geolocation because we want it to be enabled by default. However, do we need a migration for google_maps_api_key knowing that the default is to not have a key?

File: apps/qubit/modules/settings/actions/globalAction.class.php (hunk 8 above)
Comment (mcantelon) [reply apps/qubit/modules/settings/actions/globalAction.class.php]: Cool... okay I'll implement that.

Comment (mcantelon) [reply apps/qubit/modules/settings/actions/globalAction.class.php]: Okay, I've gotten rid of the Google Maps API key migration.

**Review by sevein — APPROVED** [review]

**Comment by sevein** [comment]
Squash, please!

**Comment by mcantelon** [comment]
Merged.

**Pull request closed without merging**